### PR TITLE
Errors & warnings for inappropriate RPCs

### PR DESCRIPTION
**Pull request by sjeohp**
https://github.com/paritytech/parity/issues/4605

### CODE REVIEW

File: ethcore/src/miner/miner.rs [hunk 1]
```diff
@@ -1083,6 +1083,11 @@ impl MinerService for Miner {
 		self.transaction_queue.read().last_nonce(address)
 	}
 
+	fn engine_seals_internally(&self) -> bool
+	{
```
Comment (tomusdrw) [inline ethcore/src/miner/miner.rs]: Style: keep the opening brace in the same line.

File: ethcore/src/miner/mod.rs [hunk 2]
```diff
@@ -136,6 +136,9 @@ pub trait MinerService : Send + Sync {
 	/// Called when blocks are imported to chain, updates transactions queue.
 	fn chain_new_blocks(&self, chain: &MiningBlockChainClient, imported: &[H256], invalid: &[H256], enacted: &[H256], retracted: &[H256]);
 
+	/// Engine seals internally - no external input required (ie. not an Ethash chain)
+	fn engine_seals_internally(&self) -> bool;
```
Comment (tomusdrw) [inline ethcore/src/miner/mod.rs]: I think I would call it more explicitly as: `fn can_produce_work_package(&self) -> bool`.

File: rpc/src/v1/helpers/errors.rs [hunk 3]
```diff
@@ -173,6 +174,14 @@ pub fn no_author() -> Error {
 	}
 }
 
+pub fn no_work_required() -> Error {
+	Error {
+		code: ErrorCode::ServerError(codes::NO_WORK_REQUIRED),
+		message: "External work is only required for Ethash engines.".into(),
```
Comment (tomusdrw) [inline rpc/src/v1/helpers/errors.rs]: `for Proof of Work engines`?

Comment (sjeohp) [inline rpc/src/v1/helpers/errors.rs]: Hm?  No work required for internal sealing engines, right?

**Comment by sjeohp** [comment]
[clabot:check]

File: rpc/src/v1/helpers/errors.rs (hunk 3 above)
Comment (tomusdrw) [inline rpc/src/v1/helpers/errors.rs]: Yeah, but the comment says that it's required only for `Ethash engines`. In principle we could support other engines that work the same way as Ethash (Request work package -> Submit seal).
So my point was that the comment can be made more general.

File: ethcore/src/account_provider/mod.rs [hunk 4]
```diff
@@ -64,7 +64,9 @@ pub enum SignError {
 	/// Low-level hardware device error.
 	Hardware(HardwareError),
 	/// Low-level error from store
-	SStore(SSError)
+	SStore(SSError),
+	/// Inappropriate chain
+	InappropriateChain
```
Comment (gavofyork) [inline ethcore/src/account_provider/mod.rs]: missing `,`

File: ethcore/src/account_provider/mod.rs [hunk 5]
```diff
@@ -74,6 +76,7 @@ impl fmt::Display for SignError {
 			SignError::NotFound => write!(f, "Account does not exist"),
 			SignError::Hardware(ref e) => write!(f, "{}", e),
 			SignError::SStore(ref e) => write!(f, "{}", e),
+			SignError::InappropriateChain => write!(f, "Inappropriate chain")
```
Comment (gavofyork) [inline ethcore/src/account_provider/mod.rs]: missing `,`

File: rpc/src/v1/helpers/errors.rs [hunk 6]
```diff
@@ -173,6 +174,14 @@ pub fn no_author() -> Error {
 	}
 }
 
+pub fn no_work_required() -> Error {
+	Error {
+		code: ErrorCode::ServerError(codes::NO_WORK_REQUIRED),
+		message: "External work is only required for Proof of Work engines.".into(),
+		data: None
```
Comment (gavofyork) [inline rpc/src/v1/helpers/errors.rs]: missing `,`

File: rpc/src/v1/impls/eth.rs [hunk 7]
```diff
@@ -553,6 +553,12 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM> Eth for EthClient<C, SN, S, M, EM> where
 	}
 
 	fn work(&self, no_new_work_timeout: Trailing<u64>) -> Result<Work, Error> {
+
```
Comment (gavofyork) [inline rpc/src/v1/impls/eth.rs]: superfluous (or at least non-conventional) empty line

**Comment by gavofyork** [comment]
minor style grumbles. looks good otherwise.

**Pull request merged**